src/gluonts/maybe.py

```diff
@@ -426,233 +464,430 @@ def map_or_else(
 
-        >>> Maybe(1).map_or_else(lambda n: [n], list)
+        >>> Some(1).map_or_else(lambda n: [n], list)
         [1]
-        >>> Maybe(None).map_or_else(lambda n: [n], list)
+        >>> Nothing.map_or_else(lambda n: [n], list)
         []
 
         """
-        return map_or_else(self.val, fn, factory)
 
+    @abstractmethod
     def unwrap(self) -> T:
         """
         Assert that the value is not ``None``.
 
-        >>> Maybe(1).unwrap()
+        >>> Some(1).unwrap()
         1
-        >>> Maybe(None).unwrap()
+        >>> Nothing.unwrap()
         Traceback (most recent call last):
             ...
         ValueError: Trying to unwrap `None` value.
 
         """
-        return unwrap(self.val)
 
+    @abstractmethod
     def unwrap_or(self, default: T) -> T:
         """
         Get ``val`` if it is not ``None``, or ``default`` otherwise.
 
-        >>> Maybe(1).unwrap_or(2)
+        >>> Some(1).unwrap_or(2)
         1
-        >>> Maybe(None).unwrap_or(2)
+        >>> Nothing.unwrap_or(2)
         2
 
         """
-        return unwrap_or(self.val, default)
 
+    @abstractmethod
     def unwrap_or_else(self, fn: Callable[[], T]) -> T:
         """
         Get ``val`` if it is not ``None``, or invoke ``factory`` to get a
         fallback.
 
-        >>> Maybe([1, 2, 3]).unwrap_or_else(list)
+        >>> Some([1, 2, 3]).unwrap_or_else(list)
         [1, 2, 3]
-        >>> Maybe(None).unwrap_or_else(list)
+        >>> Nothing.unwrap_or_else(list)
         []
 
         """
-        return unwrap_or_else(self.val, fn)
 
-    def and_(self, other: Optional[U]) -> Optional[U]:
+    @abstractmethod
+    def and_(self, other: OptionalOrMaybe[U]) -> Maybe[U]:
         """
         Like ``a and b`` in Python, except only considering ``None``.
 
         This is implement identical to ``unwrap_or``, but different with
         respect to types.
 
-        >>> Maybe(1).and_(2)
-        2
-        >>> Maybe(1).and_(None)
-        >>> Maybe(None).and_(2)
+        >>> Some(1).and_(2)
+        Some(2)
+        >>> Some(1).and_(None)
+        Nothing
+        >>> Nothing.and_(2)
+        Nothing
 
         """
-        return and_(self.val, other)
 
-    def __and__(self, other: Optional[U]) -> Optional[U]:
+    def __and__(self, other: OptionalOrMaybe[U]) -> Maybe[U]:
         """
         Like ``a and b`` in Python, except only considering ``None``.
 
         This is implement identical to ``unwrap_or``, but different with
         respect to types.
 
-        >>> Maybe(1) & 2
-        2
-        >>> Maybe(1) & None
-        >>> Maybe(None) & 2
+        >>> Some(1) & 2
+        Some(2)
+        >>> Some(1) & None
+        Nothing
+        >>> Nothing & 2
+        Nothing
 
         """
-        return and_(self.val, other)
+        return self.and_(other)
 
-    def and_then(self, fn: Callable[[T], Optional[U]]) -> Optional[U]:
+    @abstractmethod
+    def and_then(
+        self,
+        fn: Callable[Concatenate[T, P], U],
```

Comment:
Does `fn` always return an `Optional`? The docstring says so. I guess I’m having a hard time getting the difference with `map`

Reply:
That's the whole fun! `and_then` is `flatmap`, meaning `fn` returning an Option, makes the Monad.